qcodes/dataset/data_set.py

```diff
@@ -1333,26 +1402,31 @@ def _flush_data_to_database(self, block: bool = False) -> None:
                 argument has no effect if not using a background thread.
 
         """
+
         log.debug('Flushing to database')
+        writer_status = self._writer_status
+        if writer_status is None:
+            raise ValueError("Dataset writer has not been setup")
         if len(self._results) > 0:
             try:
+
                 self.add_results(self._results)
-                if self._bg_writer.is_alive():
+                if writer_status.write_in_background:
                     log.debug(f"Succesfully enqueued result for write thread")
                 else:
                     log.debug(f'Successfully wrote result to disk')
                 self._results = []
             except Exception as e:
-                if self._bg_writer.is_alive():
+                if writer_status.write_in_background:
                     log.warning(f"Could not enqueue result; {e}")
                 else:
                     log.warning(f'Could not commit to database; {e}')
         else:
             log.debug('No results to flush')
 
-        if self._bg_writer.is_alive() and block:
+        if writer_status.write_in_background and block:
             log.debug(f"Waiting for write queue to empty.")
-            self._data_write_queue.join()
+            writer_status.data_write_queue.join()
```

Comment:
This now bloks on all datasets being written and not just the relevant one. I think this is acceptable and makes the code a lot cleaner

Reply:
i agree. especially because one can't mix out-of-thread and main-thread writing for the same DB file.